File: quarkus-log-formatter-runtime/src/main/java/net/codinux/log/formatter/quarkus/QuarkusLogFormatterInitializer.java
package net.codinux.log.formatter.quarkus;

import net.codinux.log.formatter.quarkus.config.LogFormatterConfig;
import net.codinux.log.stacktrace.StackTraceFormatter;
import net.codinux.log.stacktrace.StackTraceFormatterOptions;
import net.codinux.log.stacktrace.StackTraceShortener;
import net.codinux.log.stacktrace.StackTraceShortenerOptions;
import io.quarkus.bootstrap.logging.QuarkusDelayedHandler;
import org.jboss.logmanager.ExtFormatter;
import org.jboss.logmanager.LogManager;
import org.jboss.logmanager.formatters.FormatStep;
import org.jboss.logmanager.formatters.FormatStep.ItemType;
import org.jboss.logmanager.formatters.Formatters;
import org.jboss.logmanager.formatters.PatternFormatter;
import org.jboss.logmanager.handlers.AsyncHandler;
import org.jboss.logmanager.handlers.ConsoleHandler;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.List;
import java.util.logging.Formatter;
import java.util.logging.Handler;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class QuarkusLogFormatterInitializer {

    public Handler initQuarkusLogFormatter() {
        return initQuarkusLogFormatter(new LogFormatterConfig());
    }

    public Handler initQuarkusLogFormatter(LogFormatterConfig config) {
        if (config.isDefault()) {
            return null; // no need to adjust ConsoleHandler, no values are set
        }

        try {
            Logger rootLogger = LogManager.getLogManager().getLogger("");

            // the first struggle is to find the ConsoleHandler which is wrapped in a QuarkusDelayedHandler and
            // in dev and test mode in another anonymous Handler and may in an AsyncHandler if
            // quarkus.log.console.async is set to true
            ConsoleHandler consoleHandler = findConsoleHandler(rootLogger);

            // the next struggle is to find the PatternFormatter, which may again is wrapped, e.g. if banner is activated
            PatternFormatter formatter = consoleHandler != null ? findPatternFormatter(consoleHandler) : null;

            if (formatter != null) {
                // and the last one is to find the field formatters like ExceptionFormatter - which again are anonymous classes
                patchFormatters(formatter, config);
            }

            return consoleHandler;
        } catch (Throwable e) {
            logError("Could not modify ConsoleHandler to format console log output", e);
            return null;
        }
    }

    private ConsoleHandler findConsoleHandler(Logger rootLogger) {
        for (Handler handler : rootLogger.getHandlers()) {
            if (handler instanceof QuarkusDelayedHandler) {
                for (Handler wrappedHandler : ((QuarkusDelayedHandler) handler).getHandlers()) {
                    ConsoleHandler found = findConsoleHandler(wrappedHandler);
                    if (found != null) {
                        return found;
                    }

                    try {
                        Class<?> wrappedHandlerClass = wrappedHandler.getClass();

                        // Quarkus creates ConsoleHandler in LoggingSetupRecorder.configureConsoleHandler() method and
                        // in dev and test mode wraps it in an anonymous Handler
                        Method enclosingMethod = wrappedHandlerClass.getEnclosingMethod();
                        if (enclosingMethod != null && enclosingMethod.getName().equals("configureConsoleHandler") &&
                                enclosingMethod.getDeclaringClass().getName().equals("io.quarkus.runtime.logging.LoggingSetupRecorder")) {
                            Field delegateField = getFirstFieldContainingOrNull(wrappedHandlerClass, "delegate");
                            if (delegateField != null) {
                                if (delegateField.trySetAccessible()) {
                                    Object delegate = delegateField.get(wrappedHandler);
                                    if (delegate instanceof ConsoleHandler) {
                                        return (ConsoleHandler) delegate;
                                    }
                                }
                            }
                        }
                    } catch (Throwable e) {
                        logError("Could not find ConsoleHandler delegate created in LoggingSetupRecorder", e);
                    }
                }
            } else {
                ConsoleHandler found = findConsoleHandler(handler);
                if (found != null) {
                    return found;
                }
            }
        }
        return null;
    }

    private ConsoleHandler findConsoleHandler(Handler handler) {
        if (handler instanceof ConsoleHandler) {
            return (ConsoleHandler) handler;
        } else if (handler instanceof AsyncHandler) {
            for (Handler inner : ((AsyncHandler) handler).getHandlers()) {
                ConsoleHandler found = findConsoleHandler(inner);
                if (found != null) {
                    return found;
                }
            }
        }

        return null;
    }

    private PatternFormatter findPatternFormatter(ConsoleHandler handler) {
        try {
            Formatter formatter = handler.getFormatter();
            if (formatter instanceof PatternFormatter) {
                return (PatternFormatter) formatter;
            } else if (formatter instanceof ExtFormatter.Delegating) { // e.g. TextBannerFormatter, which is used in case of activated banner, is derived from ExtFormatter.Delegating
                Field delegateField = getDeclaredFieldOrNull(ExtFormatter.Delegating.class, "delegate");

                if (delegateField != null) {
                    if (delegateField.trySetAccessible()) {
                        Object delegate = delegateField.get(formatter);
                        if (delegate instanceof PatternFormatter) {
                            return (PatternFormatter) delegate;
                        }
                    }
                }
            }
        } catch (Throwable e) {
            logError("Could not find PatternFormatter in ConsoleHandler", e);
        }
        return null;
    }

    private void patchFormatters(PatternFormatter formatter, LogFormatterConfig config) {
        // the step may again is wrapped, e.g. in a ColorPatternFormatter.ColorStep
        FormatStep[] originalSteps = formatter.getSteps();
        List<FormatStepInfo> stepInfos = IntStream.range(0, originalSteps.length)
                .mapToObj(i -> unwrapStep(originalSteps[i], i))
                .collect(Collectors.toList());

        for (FormatStepInfo stepInfo : stepInfos) {
            try {
                FormatStep toPatchWith = patchStepWith(stepInfo, config);
                if (toPatchWith != null) {
                    if (stepInfo.getWrappedStep() != null && stepInfo.getDelegateField() != null) {
                        stepInfo.getDelegateField().set(stepInfo.getStep(), toPatchWith);
                    } else {
                        originalSteps[stepInfo.getIndex()] = toPatchWith;
                        formatter.setSteps(originalSteps); // TODO: does it really work?
                    }
                }
            } catch (Throwable e) {
                logError("Could not patch step " + stepInfo.getType() + " " + stepInfo.getStep() + " at index " + stepInfo.getIndex(), e);
            }
        }
    }

    private FormatStep patchStepWith(FormatStepInfo stepInfo, LogFormatterConfig config) {
        if (stepInfo.getType() == ItemType.EXCEPTION_TRACE) {
            StackTraceFormatterOptions options = new StackTraceFormatterOptions.Builder()
                    .rootCauseFirst(config.isRootCauseFirst())
                    .maxStackTraceStringLength(config.getMaxStackTraceStringLength())
                    .build();
            StackTraceShortener shortener = new StackTraceShortener(new StackTraceShortenerOptions(
                    config.getMaxFramesPerThrowable(), config.getMaxNestedThrowables()));

            return new ExceptionFormatStep(new StackTraceFormatter(options, shortener), options.getLineSeparator());
        }

        return null;
    }

    private FormatStepInfo unwrapStep(FormatStep step, int index) {
        try {
            Field delegateField = getDeclaredFieldOrNull(step.getClass(), "delegate");
            if (delegateField != null) {
                if (delegateField.trySetAccessible()) {
                    Object delegate = delegateField.get(step);
                    if (delegate instanceof FormatStep) {
                        return new FormatStepInfo(step, getStepType((FormatStep) delegate), index, (FormatStep) delegate, delegateField);
                    }
                }
            }
        } catch (Throwable e) {
            logError("Could not unwrap delegate of step " + step + " at index " + index, e);
        }

        return new FormatStepInfo(step, getStepType(step), index, null, null);
    }

    private ItemType getStepType(FormatStep step) {
        try {
            Class stepClass = step.getClass();

            Method getItemTypeMethod = getDeclaredMethodOrNull(stepClass, "getItemType");
            if (getItemTypeMethod != null && getItemTypeMethod.getReturnType() == ItemType.class && getItemTypeMethod.getParameterCount() == 0) {
                getItemTypeMethod.trySetAccessible();
                return (ItemType) getItemTypeMethod.invoke(step);
            }

            Method enclosing = stepClass.getEnclosingMethod();
            if (enclosing != null && enclosing.getDeclaringClass() == Formatters.class) {
                if (enclosing.getName().equals("exceptionFormatStep")) {
                    return ItemType.EXCEPTION_TRACE;
                }
            }
        } catch (Throwable e) {
            logError("Could not get step type of " + step, e);
        }
        return null;
    }


    private void logError(String message, Throwable error) {
        Logger.getLogger(QuarkusLogFormatterInitializer.class.getName()).log(Level.SEVERE, message, error);
    }


    private Method getDeclaredMethodOrNull(Class<?> clazz, String name) {
        for (Method method : clazz.getDeclaredMethods()) {
            if (method.getName().equals(name)) {
                return method;
            }
        }

        return null;
    }

    private Field getDeclaredFieldOrNull(Class<?> clazz, String name) {
        for (Field field : clazz.getDeclaredFields()) {
            if (field.getName().equals(name)) {
                return field;
            }
        }

        return null;
    }

    private Field getFirstFieldContainingOrNull(Class<?> clazz, String substring) {
        for (Field field : clazz.getDeclaredFields()) {
            if (field.getName().contains(substring)) {
                return field;
            }
        }

        return null;
    }

}

File: quarkus-log-formatter-logic/src/main/kotlin/net/codinux/log/formatter/quarkus/QuarkusLogFormatterInitializer.kt
package net.codinux.log.formatter.quarkus

import io.quarkus.bootstrap.logging.QuarkusDelayedHandler
import net.codinux.log.formatter.quarkus.config.LogFormatterConfig
import net.codinux.log.stacktrace.StackTraceFormatter
import net.codinux.log.stacktrace.StackTraceFormatterOptions
import net.codinux.log.stacktrace.StackTraceShortener
import net.codinux.log.stacktrace.StackTraceShortenerOptions
import org.jboss.logmanager.ExtFormatter
import org.jboss.logmanager.LogManager
import org.jboss.logmanager.formatters.FormatStep
import org.jboss.logmanager.formatters.FormatStep.ItemType
import org.jboss.logmanager.formatters.Formatters
import org.jboss.logmanager.formatters.PatternFormatter
import org.jboss.logmanager.handlers.AsyncHandler
import org.jboss.logmanager.handlers.ConsoleHandler
import java.util.logging.*

class QuarkusLogFormatterInitializer {

    @JvmOverloads
    fun initQuarkusLogFormatter(config: LogFormatterConfig = LogFormatterConfig()): Handler? {
        if (config.isDefault) {
            return null // no need to adjust ConsoleHandler, no values are set
        }

        try {
            val rootLogger = LogManager.getLogManager().getLogger("")

            // the first struggle is to find the ConsoleHandler which is wrapped in a QuarkusDelayedHandler and
            // in dev and test mode in another anonymous Handler and may in an AsyncHandler if
            // quarkus.log.console.async is set to true
            val consoleHandler = findConsoleHandler(rootLogger)

            // the next struggle is to find the PatternFormatter, which may again is wrapped, e.g. if banner is activated
            val formatter = consoleHandler?.let { findPatternFormatter(it) }

            // and the last one is to find the field formatters like ExceptionFormatter - which again are anonymous classes
            formatter?.let { patchFormatters(it, config) }

            return consoleHandler
        } catch (e: Throwable) {
            logError("Could not modify ConsoleHandler to format console log output", e)
            return null
        }
    }


    private fun findConsoleHandler(rootLogger: Logger): ConsoleHandler? {
        rootLogger.handlers.forEach { handler ->
            val handlerClass = handler.javaClass

            if (handler is QuarkusDelayedHandler) {
                val wrappedHandlers = handler.handlers
                wrappedHandlers.forEach { wrappedHandler ->
                    findConsoleHandler(wrappedHandler)?.let {
                        return it
                    }

                    try {
                        val wrappedHandlerClass = wrappedHandler.javaClass

                        // Quarkus creates ConsoleHandler in LoggingSetupRecorder.configureConsoleHandler() method and
                        // in dev and test mode wraps it in an anonymous Handler
                        val enclosingMethod = wrappedHandlerClass.enclosingMethod
                        if (enclosingMethod != null && enclosingMethod.name == "configureConsoleHandler" &&
                            enclosingMethod.declaringClass.name == "io.quarkus.runtime.logging.LoggingSetupRecorder") {
                            val delegateField = wrappedHandlerClass.declaredFields.firstOrNull { it.name.contains("delegate") }
                            if (delegateField != null) {
                                if (delegateField.trySetAccessible()) {
                                    val delegate = delegateField.get(wrappedHandler)
                                    if (delegate is ConsoleHandler) {
                                        return delegate
                                    }
                                }
                            }
                        }
                    } catch (e: Throwable) {
                        logError("Could not find ConsoleHandler delegate created in LoggingSetupRecorder", e)
                    }
                }
            } else {
                findConsoleHandler(handler)?.let {
                    return it
                }
            }
        }

        return null
    }

    private fun findConsoleHandler(handler: Handler): ConsoleHandler? =
        if (handler is ConsoleHandler) {
            handler
        } else if (handler is AsyncHandler) {
            handler.handlers.firstNotNullOfOrNull { findConsoleHandler(it) }
        } else {
            null
        }


    private fun findPatternFormatter(handler: ConsoleHandler): PatternFormatter? {
        try {
            val formatter = handler.formatter
            if (formatter is PatternFormatter) {
                return formatter
            } else if (formatter is ExtFormatter.Delegating) { // e.g. TextBannerFormatter, which is used in case of activated banner, is derived from ExtFormatter.Delegating
                val delegatingFormatterClass = ExtFormatter.Delegating::class.java
                val delegateField = delegatingFormatterClass.declaredFields.firstOrNull { it.name == "delegate" }
                if (delegateField != null) {
                    if (delegateField.trySetAccessible()) {
                        val delegate = delegateField.get(formatter) as? Formatter
                        if (delegate is PatternFormatter) {
                            return delegate
                        }
                    }
                }
            }
        } catch (e: Throwable) {
            logError("Could not find PatternFormatter in ConsoleHandler", e)
        }

        return null
    }


    private fun patchFormatters(formatter: PatternFormatter, config: LogFormatterConfig) {
        // the step may again is wrapped, e.g. in a ColorPatternFormatter.ColorStep
        val steps = formatter.steps.mapIndexed { index, step -> unwrapStep(step, index) }

        steps.forEach { stepInfo ->
            try {
                patchStepWith(stepInfo, config)?.let { toPatchWith ->
                    if (stepInfo.wrappedStep != null && stepInfo.delegateField != null) { // then replace the wrapped FormatStep
                        stepInfo.delegateField.set(stepInfo.step, toPatchWith)
                    } else {
                        formatter.steps = formatter.steps.apply {
                            this[stepInfo.index] = toPatchWith
                        }
                    }
                }
            } catch (e: Throwable) {
                logError("Could not patch step ${stepInfo.type} ${stepInfo.step} at index ${stepInfo.index}", e)
            }
        }
    }

    private fun patchStepWith(stepInfo: FormatStepInfo, config: LogFormatterConfig): FormatStep? =
        if (stepInfo.type == ItemType.EXCEPTION_TRACE) {
            val options = StackTraceFormatterOptions(rootCauseFirst = config.rootCauseFirst, maxStackTraceStringLength = config.maxStackTraceStringLength)
            val shortener = StackTraceShortener(StackTraceShortenerOptions(config.maxFramesPerThrowable, config.maxNestedThrowables))
            ExceptionFormatStep(StackTraceFormatter(options, shortener), options.lineSeparator)
        } else {
            null
        }

    private fun unwrapStep(step: FormatStep, index: Int): FormatStepInfo {
        try {
            val delegateField = step.javaClass.declaredFields.firstOrNull { it.name == "delegate" }
            if (delegateField != null) {
                if (delegateField.trySetAccessible()) {
                    val delegate = delegateField.get(step)
                    if (delegate is FormatStep) {
                        return FormatStepInfo(step, getStepType(delegate), index, delegate, delegateField)
                    }
                }
            }
        } catch (e: Throwable) {
            logError("Could not unwrap delegate of step $step at index $index", e)
        }

        return FormatStepInfo(step, getStepType(step), index, null, null)
    }

    private fun getStepType(step: FormatStep): ItemType? {
        try {
            val stepClass = step.javaClass

            val getItemTypeMethod = stepClass.declaredMethods.firstOrNull { it.name == "getItemType" && it.parameterTypes.isEmpty() && it.returnType == ItemType::class.java }
            if (getItemTypeMethod != null) {
                getItemTypeMethod.trySetAccessible()
                return getItemTypeMethod.invoke(step) as? ItemType
            }

            val enclosingMethod = stepClass.enclosingMethod
            if (enclosingMethod != null && enclosingMethod.declaringClass == Formatters::class.java) {
                if (enclosingMethod.name == "exceptionFormatStep") {
                    return ItemType.EXCEPTION_TRACE
                }
            }
        } catch (e: Throwable) {
            logError("Could not get step type of $step", e)
        }

        return null
    }


    private fun logError(message: String, error: Throwable?) {
        Logger.getLogger(QuarkusLogFormatterInitializer::class.java.name).log(Level.SEVERE, message, error)
    }

}